wrappers/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,542 @@
+
+func MechanisFromString(mech string) (uint, error) {
+	switch mech {
+	case "CKM_RSA_PKCS":
+		return pkcs11.CKM_RSA_PKCS, nil
+	case "CKM_AES_CBC_PAD":
+		return pkcs11.CKM_AES_CBC_PAD, nil
+	default:
+		return 0, fmt.Errorf("unsupported mechanism: %s", mech)
+	}
+}
+
+type pkcs11KMS struct {
+	// standard PKCS11 configuration options
+	slot      uint
+	pin       string
+	module    string
+	keyId     string
+	label     string
+	mechanism uint
+}
+
+// EncryptDEK uses the PKCS11 encrypt operation to encrypt the DEK.
+func (kms *pkcs11KMS) EncryptDEK(ctx context.Context, plainDEK []byte) ([]byte, error) {
+	p := pkcs11.New(kms.module)
+	err := p.Initialize()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize PKCS11: %w", err)
+	}
+
+	defer p.Destroy()
+	defer p.Finalize()
+
+	session, err := p.OpenSession(kms.slot, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open session: %w", err)
+	}
+	defer p.CloseSession(session)
+
+	err = p.Login(session, pkcs11.CKU_USER, kms.pin)
+	if err != nil {
+		return nil, fmt.Errorf("failed to login: %w", err)
+	}
+	defer p.Logout(session)
+
+	keyIdBytes, err := hex.DecodeString(kms.keyId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode key id: %w", err)
+	}
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_ID, keyIdBytes),
+		pkcs11.NewAttribute(pkcs11.CKA_ENCRYPT, true),
+	}
+	if kms.label != "" {
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_LABEL, kms.label))
+	}
+	if kms.mechanism != 0 {
+		keyTypeString, err := GetKeyTypeFromMech(kms.mechanism)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get key type from mechanism: %s", err)
+		}
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, keyTypeString))
+	}
+	if err := p.FindObjectsInit(session, template); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjectsInit: %s", err)
+	}
+	obj, _, err := p.FindObjects(session, 2)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjects: %s", err)
+	}
+	if err := p.FindObjectsFinal(session); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjectsFinal: %s", err)
+	}
+
+	if len(obj) != 1 {
+		return nil, fmt.Errorf("expected 1 object, got %d", len(obj))
+	}
+	key := obj[0]
+
+	template = []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, nil),
+	}
+	attr, err := p.GetAttributeValue(session, pkcs11.ObjectHandle(key), template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 GetAttributeValue: %s", err)
+	}
+
+	attrMap := GetAttributesMap(attr)
+	keyType := GetValueAsInt(attrMap[pkcs11.CKA_KEY_TYPE])
+
+	mechanism := uint(0)
+	switch keyType {
+	case pkcs11.CKK_AES:
+		if kms.mechanism != 0 {
+			mechanism = kms.mechanism
+		} else {
+			mechanism = pkcs11.CKM_AES_CBC_PAD
+		}
+	case pkcs11.CKK_RSA:
+		if kms.mechanism != 0 {
+			mechanism = kms.mechanism
+		} else {
+			mechanism = pkcs11.CKM_RSA_PKCS
```

Comment:
```suggestion
			mechanism = pkcs11.CKM_RSA_PKCS_OAEP
```

Vault [prefers this](https://developer.hashicorp.com/vault/docs/configuration/seal/pkcs11#mechanism).